dialer.go

```diff
@@ -68,175 +81,168 @@ func NewDialer(opts *DialerOpts) Dialer {
-	liveSessions          chan sessionIntf
-	emaRTT                *ema.EMA
+	// We create another background session with a shorter dial timeout to ensure liveness in the case of network
+	// disruptions.
+	d.pendingSessions <- &sessionConfig{
+		name:         "liveness to " + d.name,
+		dialTimeout:  5 * time.Second,
+		sleepOnError: 1 * time.Second,
+	}
+	ops.Go(d.maintainTCPConnections)
+	return d
 }
 
-func (d *dialer) Dial(dial DialFN) (net.Conn, error) {
-	return d.DialContext(context.Background(), dial)
+// maintainTCPConnections maintains background TCP connection(s) and associated lampshade session(s)
+func (d *dialer) maintainTCPConnections() {
+	for sc := range d.pendingSessions {
+		d.trySession(sc)
+	}
 }
 
-func (d *dialer) DialContext(ctx context.Context, dial DialFN) (net.Conn, error) {
-	s, err := d.getOrCreateSession(ctx, dial)
+func (d *dialer) trySession(sc *sessionConfig) {
+	start := time.Now()
+	s, err := d.startSession(sc)
 	if err != nil {
-		return nil, err
+		log.Debugf("Error starting session '%v': %v", sc.name, err.Error())
+		time.Sleep(sc.sleepOnError)
+		d.pendingSessions <- sc
+	} else {
+		log.Debugf("Created session in %v to %#v", time.Since(start), sc)
+		d.recycleSession(s)
 	}
-	c := s.CreateStream()
-	d.returnSession(s)
-	return c, nil
 }
 
-func (d *dialer) getNumLivePending() int {
-	d.muNumLivePending.Lock()
-	numLivePending := d.numLive + d.numPending
-	d.muNumLivePending.Unlock()
-	return numLivePending
+func (d *dialer) recycleSession(s sessionIntf) {
+	// We now have a new or established TCP connection/session. At this point two things can happen:
+	// 1) The connection can be closed for any reason, in which case we want to request a new one
+	// 2) A dialer can request the session. In that case, it will retrieve the session.
+	select {
+	case <-s.getCloseCh():
+		log.Debugf("Session closed before requested. Requesting new session: %#v", s.getSessionConfig())
+		d.pendingSessions <- s.getSessionConfig()
+	case <-d.sessionRequests:
+		d.liveSessions <- s
+	}
 }
 
-func (d *dialer) getOrCreateSession(ctx context.Context, dial DialFN) (sessionIntf, error) {
-	newSession := func(cap int) {
-		d.muNumLivePending.Lock()
-		if d.numLive+d.numPending >= cap {
-			d.muNumLivePending.Unlock()
-			return
-		}
-		d.numPending++
-		d.muNumLivePending.Unlock()
-		go func() {
-			s, err := d.startSession(dial)
-			d.muNumLivePending.Lock()
-			d.numPending--
-			if err != nil {
-				d.muNumLivePending.Unlock()
-				return
-			}
-			d.numLive++
-			d.muNumLivePending.Unlock()
-			d.liveSessions <- s
-		}()
+func (d *dialer) Dial() (net.Conn, error) {
+	return d.DialContext(context.Background())
+}
+
+func (d *dialer) DialContext(ctx context.Context) (net.Conn, error) {
+	s, err := d.getSession(ctx)
+	if err != nil {
+		return nil, err
 	}
+	c := s.createStream(ctx)
+	go d.recycleSession(s)
+
+	return c, nil
+}
+
+func (d *dialer) getSession(ctx context.Context) (sessionIntf, error) {
+	d.sessionRequests <- true
```

Comment:
Some goroutines block here
```
48 @ 0x4031fef 0x4007543 0x4007519 0x4007295 0x4742a0e 0x4742884 0x4961a3e 0x495d15d 0x4961c7c 0x495be6c 0x4946d9c 0x49467e0 0x4945adf 0x495f437 0x44812da 0x44876a1 0x4497e31 0x4061081
#	0x4742a0d	github.com/getlantern/lampshade.(*dialer).getSession+0x7d		/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190910214900-cfa0ddb2100f/dialer.go:184
#	0x4742883	github.com/getlantern/lampshade.(*dialer).DialContext+0x53		/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190910214900-cfa0ddb2100f/dialer.go:173
...
#	0x4497e30	net/http.(*Transport).getConn.func4+0xd0				/usr/local/Cellar/go/1.12.6/libexec/src/net/http/transport.go:1015
...

10 @ 0x4031fef 0x4007543 0x4007519 0x4007295 0x4742a0e 0x4742884 0x4961a3e 0x495d15d 0x4961c7c 0x495be6c 0x4946d9c 0x49467e0 0x4945adf 0x471ae03 0x471a4a5 0x4719c44 0x4a91b43 0x4a92632 0x4a8c736 0x4a8be97 0x4758152 0x475ad46 0x4759743 0x475878f 0x4a8e79d 0x4061081
#	0x4742a0d	github.com/getlantern/lampshade.(*dialer).getSession+0x7d			/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190910214900-cfa0ddb2100f/dialer.go:184
#	0x4742883	github.com/getlantern/lampshade.(*dialer).DialContext+0x53			/Users/joesis/go/pkg/mod/github.com/getlantern/lampshade@v0.0.0-20190910214900-cfa0ddb2100f/dialer.go:173
...
#	0x4a8e79c	github.com/getlantern/flashlight/client.(*Client).handle+0x19c			/Users/joesis/workspace/go/src/github.com/getlantern/flashlight/client/handler.go:33
```

Reply:
Okay so if `s.getSession()` times out, `recycleSession` won't be called. If it happened for 10 times, the `sessionRequests` would be full, cause all later `s.getSession()` timing out as well. This patch fixed it for me, though the role of `sessionRequests` is not very obvious at first sight.

```
diff --git a/dialer.go b/dialer.go
index 1004025..f4621e2 100644
--- a/dialer.go
+++ b/dialer.go
@@ -171,13 +171,11 @@ func (d *dialer) Dial() (net.Conn, error) {

 func (d *dialer) DialContext(ctx context.Context) (net.Conn, error) {
        s, err := d.getSession(ctx)
+       go d.recycleSession(s)
        if err != nil {
                return nil, err
        }
-       c := s.createStream(ctx)
-       go d.recycleSession(s)
-
-       return c, nil
+       return s.createStream(ctx), nil
 }
```

Reply:
OK fixed! Another great catch! There might be some simpler way, but `sessionRequests` is just intended to trigger `recycleSessions` into knowing there's an active request, and that it's therefore safe to put the current session in `liveSessions`. That way `liveSessions` never has almost never contains closed sessions, and the only time it can have a closed session is if a session is closed between `recycleSessions` putting it in `liveSessions` and `getSession` pulling it out of `liveSessions`, but that will be caught by the `isClosed` check.

Reply:
Ha I realized I introduced another bug which could call `recycleSession` with an invalid session, but luckily you fixed it.